### PR TITLE
Add the beginnings of a performance optimization guide.

### PR DESCRIPTION
**Pull request by saeta**

### CODE REVIEW

File: Sources/PenguinStructures/Documentation/Guides/Perf.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    command line generated in the previous step. After running the compiler, we can inspect the SIL
+    by opening the file `PenguinStructures.sil` with our favorite text editor.
+
+ 5. **Making changes**: As you tweak the program, just re-run the command from the previous step,
+    and it will output the updated SIL into the same file. This makes it very easy to see the effect
+    different changes to the source has on the resulting program. Beware, if you change files in
+    other modules, you may need to re-build with SwiftPM and re-generate the command line to dump
+    IRs.
+
+## Perennial Performance Pitfalls
```
[the remark this answers is not in the document]
Comment (saeta) [reply Sources/PenguinStructures/Documentation/Guides/Perf.md:219]: +1 Thanks! I've added a section for escaping closures, as well as `-cross-module-optimization` in a new "don't forget to compile with optimization section".

File: Sources/PenguinStructures/Documentation/Guides/Perf.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+ 1. **Write a benchmark**: The current recommendation is to use the [Swift
+    Benchmark](https://github.com/google/swift-benchmark) library, as it has a nice suite of
+    features. Ensure it is representative of the program / subroutine you're trying to optimize.
+    Try to write your benchmarks so they run quickly to enable a fast iteratation cycle. (Note: use
+    Swift Benchmark's `--filter` flag to run only the relevant benchmarks you're trying to optimze.)
+
+    When developing with SwiftPM, we've found that defining an (unexported) `Benchmarks` executable
+    target that aggregates all your benchmarks across your project into a single binary works well.
+
+ 2. **Start a spreadsheet & gather baselines**: Keep track of what changes you're making, and how
```
[the remark this answers is not in the document]
Comment (saeta) [reply Sources/PenguinStructures/Documentation/Guides/Perf.md:21]: 👍

File: Sources/PenguinStructures/Documentation/Guides/Perf.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+## Preparation
+
+> Performance optimization is incredibly subtle. Both hardware and the Swift compiler are very
+> sophisticated; and both have non-obvious performance characteristics. By doing some preparation,
+> you can remove a lot of guesswork and save yourself time.
+
+ 1. **Write a benchmark**: The current recommendation is to use the [Swift
+    Benchmark](https://github.com/google/swift-benchmark) library, as it has a nice suite of
+    features. Ensure it is representative of the program / subroutine you're trying to optimize.
+    Try to write your benchmarks so they run quickly to enable a fast iteratation cycle. (Note: use
```
[the remark this answers is not in the document]
Comment (saeta) [reply Sources/PenguinStructures/Documentation/Guides/Perf.md:15]: Great catch. Thank you!

**Comment by saeta** [comment]
> One area that I'd emphasize early on, because it's a very common source of confusion, is that Swift's performance varies (wildly, in some cases) between unoptimized builds (the default for SwiftPM) and release / optimized builds. That would be something I'd highlight and make explicit. It's implicitly there in the command-line arguments in your examples, but I've seen many people trip over this.

Excellent point. I've added this as the first point in the P3 section, and augmented it with `-cross-module-optimization` as well.

> I have my own workflow that's heavily reliant on Instruments on the Mac, would it be useful to add that later once this is in?

@BradLarson YES PLEASE! :-)

File: Sources/PenguinStructures/Documentation/Guides/Perf.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (shabalind) [reply Sources/PenguinStructures/Documentation/Guides/Perf.md:219]: Great, thanks! 👍

**Pull request merged**